src/ZEO/ClientStorage.py

```diff
@@ -1388,6 +1389,15 @@ def verify_cache(self, server):
         for oid, tid in self._cache.contents():
             server.verify(oid, tid)
         server.endZeoVerify()
+
+        with self._lock:
+            if server_tid > self._cache.getLastTid():
+                # We verified the cache, and got no new invalidations
+                # while doing so.  The records in the cache are valid,
+                # in that invalid current records were invalidated,
+                # but the last tid is wrong.  Let's fix it:
+                self._cache.setLastTid(server_tid)
```

Comment:
I see that `finish_verification` acquires and releases the lock, so this should close the possible race, and I don't see any possibility of deadlock.